{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge patient_to_mutation output files and upload it to MySQL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list = ['../output/patient_to_mutation_'+str(i)+'.csv' for i in [1,2,3,4,5,6,8,9,15,16]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "df = pd.DataFrame()\n",
    "for index in range(len(file_list)):  \n",
    "    data = pd.read_csv(file_list[index])\n",
    "    try:\n",
    "        data = data.drop(\"Unnamed: 0\", axis =1)\n",
    "        data = data.drop(\"Unnamed: 0.1\", axis =1)\n",
    "        data = data.drop(\"Unnamed: 0.1.1\", axis =1)\n",
    "    except:\n",
    "        a=1\n",
    "    df = df.append(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation_id</th>\n",
       "      <th>Allele_Fraction</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Compound_Heterozygous</th>\n",
       "      <th>Curated</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Lab_UPN</th>\n",
       "      <th>Read_Depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:10566230_C--&gt;A</td>\n",
       "      <td>2.67</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>CO22348-0257</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:10566272_C--&gt;T</td>\n",
       "      <td>51.98</td>\n",
       "      <td>Benign</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>CO22348-0257</td>\n",
       "      <td>177.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:45797154_G--&gt;A</td>\n",
       "      <td>2.02</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>CO22348-0257</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:45797505_C--&gt;G</td>\n",
       "      <td>47.12</td>\n",
       "      <td>Benign</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>CO22348-0257</td>\n",
       "      <td>104.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:45797951_G--&gt;X</td>\n",
       "      <td>0.81</td>\n",
       "      <td>Likely Pathogenic</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>CO22348-0257</td>\n",
       "      <td>248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320350</th>\n",
       "      <td>chrX:154348330_T--&gt;A</td>\n",
       "      <td>2.08</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>BCA46684</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320351</th>\n",
       "      <td>chrX:154348389_G--&gt;A</td>\n",
       "      <td>2.27</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>BCA46684</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320352</th>\n",
       "      <td>chrY:21154323_G--&gt;A</td>\n",
       "      <td>99.31</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Hemi</td>\n",
       "      <td>BCA46684</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320353</th>\n",
       "      <td>chrY:21154426_G--&gt;A</td>\n",
       "      <td>37.70</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Het</td>\n",
       "      <td>BCA46684</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320354</th>\n",
       "      <td>chrY:21154466_T--&gt;A</td>\n",
       "      <td>98.98</td>\n",
       "      <td>Benign</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Hemi</td>\n",
       "      <td>BCA46684</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2063521 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Mutation_id  Allele_Fraction          Classification  \\\n",
       "0        chr1:10566230_C-->A             2.67  Uncertain Significance   \n",
       "1        chr1:10566272_C-->T            51.98                  Benign   \n",
       "2        chr1:45797154_G-->A             2.02  Uncertain Significance   \n",
       "3        chr1:45797505_C-->G            47.12                  Benign   \n",
       "4        chr1:45797951_G-->X             0.81       Likely Pathogenic   \n",
       "...                      ...              ...                     ...   \n",
       "320350  chrX:154348330_T-->A             2.08  Uncertain Significance   \n",
       "320351  chrX:154348389_G-->A             2.27  Uncertain Significance   \n",
       "320352   chrY:21154323_G-->A            99.31  Uncertain Significance   \n",
       "320353   chrY:21154426_G-->A            37.70  Uncertain Significance   \n",
       "320354   chrY:21154466_T-->A            98.98                  Benign   \n",
       "\n",
       "       Compound_Heterozygous Curated Genotype       Lab_UPN  Read_Depth  \n",
       "0                         No      No      Het  CO22348-0257       150.0  \n",
       "1                         No      No      Het  CO22348-0257       177.0  \n",
       "2                         No      No      Het  CO22348-0257        99.0  \n",
       "3                         No      No      Het  CO22348-0257       104.0  \n",
       "4                         No      No      Het  CO22348-0257       248.0  \n",
       "...                      ...     ...      ...           ...         ...  \n",
       "320350                    No      No      Het      BCA46684        96.0  \n",
       "320351                    No      No      Het      BCA46684        88.0  \n",
       "320352                    No      No     Hemi      BCA46684       144.0  \n",
       "320353                    No      No      Het      BCA46684        61.0  \n",
       "320354                    No      No     Hemi      BCA46684        98.0  \n",
       "\n",
       "[2063521 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../output/patient_to_mutation_merged.csv', index=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import dataframe into MySQL\n",
    "import sqlalchemy\n",
    "database_username = 'root'\n",
    "database_password = 'Kjy11422@'\n",
    "database_ip       = '127.0.0.1'\n",
    "database_name     = 'Genomics'\n",
    "database_connection = sqlalchemy.create_engine('mysql+mysqlconnector://{0}:{1}@{2}/{3}'.\n",
    "                                               format(database_username, database_password, \n",
    "                                                      database_ip, database_name))\n",
    "df.to_sql(con=database_connection, name='Genomics_merged', if_exists='replace')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          CO22348-0257\n",
       "1364       CO23880-0736\n",
       "2650      UPR12664-0780\n",
       "3806      UPR12657-0770\n",
       "4847       CO23868-1084\n",
       "              ...      \n",
       "312183         BCA47789\n",
       "313857    GDL27234-0670\n",
       "315471         BCA47729\n",
       "317108         BCA45714\n",
       "318685         BCA46684\n",
       "Name: Lab_UPN, Length: 1443, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Lab_UPN\"].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
