{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from flask import Flask, render_template, request\n",
    "import json\n",
    "\n",
    "app = Flask(__name__)\n",
    "\n",
    "\n",
    "\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.conf import SparkConf\n",
    "from pyspark.sql import SQLContext\n",
    "from pyspark import SparkContext\n",
    "my_spark = SparkSession \\\n",
    "    .builder \\\n",
    "    .appName(\"\") \\\n",
    "    .config(\"spark.mongodb.input.uri=mongodb://127.0.0.1/Patient_Registry.Patient_Registry\") \\\n",
    "    .config(\"spark.mongodb.output.uri=mongodb://127.0.0.1/Patient_Registry.Patient_Registry\") \\\n",
    "    .config('spark.jars.packages', 'org.mongodb.spark:mongo-spark-connector_2.12:3.0.0') \\\n",
    "    .config('spark.driver.extraClassPath', '/Users/jaeyoungkim/Desktop/usc/DSCI - 551/project/code/jars/mysql-connector-java-8.0.11.jar')\\\n",
    "    .getOrCreate()\n",
    "\n",
    "hostname = \"127.0.0.1\"\n",
    "dbname = 'Genomics'\n",
    "jdbcPort = 3306\n",
    "username = 'root'\n",
    "password = 'Kjy11422@'\n",
    "jdbc_url = \"jdbc:mysql://{0}:{1}/{2}?user={3}&password={4}&characterEncoding=UTF-8&serverTimezone=UTC\".format(hostname,jdbcPort,dbname,username,password)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app \"__main__\" (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [27/Nov/2020 14:45:13] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2020 14:45:13] \"\u001b[33mGET /static/js/extention/choices.js HTTP/1.1\u001b[0m\" 404 -\n",
      "127.0.0.1 - - [27/Nov/2020 14:45:13] \"\u001b[37mGET /static/css/main.css HTTP/1.1\u001b[0m\" 200 -\n",
      "127.0.0.1 - - [27/Nov/2020 14:45:13] \"\u001b[33mGET /static/js/extention/choices.js HTTP/1.1\u001b[0m\" 404 -\n",
      "127.0.0.1 - - [27/Nov/2020 14:45:13] \"\u001b[37mGET /static/images/gene.png HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "MUTYH\n",
      "+----------+--------------------+-----------+-------------------+--------+--------------------+----------------+-------------+--------------------+--------------------+----------------+--------------+--------------------+\n",
      "|Chromosome|         Gene Region|Gene Symbol|        Mutation_ID|Position|     Protein Variant|Reference Allele|Sample Allele|       Transcript ID|  Transcript Variant|Variant Findings|Variation Type|                 _id|\n",
      "+----------+--------------------+-----------+-------------------+--------+--------------------+----------------+-------------+--------------------+--------------------+----------------+--------------+--------------------+\n",
      "|         1|        ncRNA; 3'UTR|      MUTYH|chr1:45794974_G-->A|45794974|               [nan]|               G|            A|[ENST00000448481....|[n.1822C>T, n.200...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45794979_C-->A|45794979|[p.*217L, p.*536L...|               C|            A|[ENST00000448481....|[c.1598G>T, c.161...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|               null|45794983|[p.Q536fs*?, p.Q5...|               G|         null|[ENST00000448481....|[c.1636delC, c.12...|            null|      Deletion|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45794985_G-->A|45794985|[p.A521V, p.A531V...|               G|            A|[ENST00000448481....|[c.644C>T, c.1634...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795003_T-->C|45795003|[p.H542R, p.H209R...|               T|            C|[ENST00000448481....|[c.1586A>G, c.161...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795005_T-->C|45795005|[p.A524A, p.A513A...|               T|            C|[ENST00000448481....|[c.1263A>G, c.158...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795013_T-->A|45795013|[p.T525S, p.T536S...|               T|            A|[ENST00000448481....|[c.*23A>T, n.1783...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795014_G-->A|45795014|[p.S538S, p.S418S...|               G|            A|[ENST00000448481....|[c.1572C>T, c.157...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795015_G-->A|45795015|[p.S525F, p.S418F...|               G|            A|[ENST00000448481....|[c.1574C>T, c.160...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795015_G-->T|45795015|[p.S418Y, p.S524Y...|               G|            T|[ENST00000448481....|[c.1562C>A, c.153...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795016_A-->C|45795016|[p.S521A, p.S510A...|               A|            C|[ENST00000448481....|[c.1528T>G, c.161...|             3.0|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795021_T-->A|45795021|[p.H203L, p.H523L...|               T|            A|[ENST00000448481....|[c.1523A>T, c.160...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|Exonic; ncRNA; 3'UTR|      MUTYH|chr1:45795031_A-->G|45795031|[p.F390L, p.F200L...|               A|            G|[ENST00000448481....|[n.1951T>C, c.159...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795041_C-->G|45795041|[p.L516L, p.L515L...|               C|            G|[ENST00000448481....|[n.1941G>C, c.158...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795049_G-->T|45795049|[p.Q513K, p.Q524K...|               G|            T|[ENST00000448481....|[c.1498C>A, c.153...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|               null|45795052|[p.Q512fs*?, p.Q5...|               G|         null|[ENST00000448481....|[c.1216delC, c.15...|            null|      Deletion|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795061_G-->A|45795061|[p.R506C, p.R509C...|               G|            A|[ENST00000448481....|[n.1735C>T, c.148...|             8.0|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795066_T-->C|45795066|[p.K494R, p.K493R...|               T|            C|[ENST00000448481....|[c.1478A>G, c.156...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795074_A-->T|45795074|[p.S518R, p.S504R...|               A|            T|[ENST00000448481....|[c.1545T>A, c.555...|            null|           SNV|[5fa2d3fbcc769b52...|\n",
      "|         1|       Exonic; ncRNA|      MUTYH|chr1:45795081_G-->A|45795081|[p.P183L, p.P489L...|               G|            A|[ENST00000448481....|[c.1547C>T, c.146...|            21.0|           SNV|[5fa2d3fbcc769b52...|\n",
      "+----------+--------------------+-----------+-------------------+--------+--------------------+----------------+-------------+--------------------+--------------------+----------------+--------------+--------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Nov/2020 14:46:02] \"\u001b[37mPOST /test/Pro_var HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    }
   ],
   "source": [
    "\n",
    "@app.route(\"/\")\n",
    "def index():\n",
    "    return render_template('index.html')\n",
    "\n",
    "\n",
    "\n",
    "@app.route(\"/test/Mut_ID\", methods = ['Post'])\n",
    "def Mutation_ID():\n",
    "    Mutation_ID = request.form['l']\n",
    "#     print(Mutation_ID)\n",
    "    pipeline = \"{'$match' : {Mutation_ID:'\"+Mutation_ID+\"'}}\"  ## mutationID location\n",
    "    df01 = my_spark.read\\\n",
    "        .format(\"mongo\")\\\n",
    "        .option(\"database\",\"Mut_info\")\\\n",
    "        .option(\"pipeline\", pipeline)\\\n",
    "        .option(\"collection\",\"Mutation_Information_Merged\")\\\n",
    "        .load().limit(1)\n",
    "\n",
    "    res = df01.toJSON().collect()\n",
    "    res = json.loads(res[0])\n",
    "#     print(res)\n",
    "    tmp=list(res.keys())\n",
    "#     print(tmp)\n",
    "    return render_template('mut_inf_result.html', parameter=res, tmp = tmp, par = Mutation_ID)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "@app.route(\"/test/<parameter>\")\n",
    "def MutationLink(parameter):\n",
    "    pipeline = \"{'$match' : {Mutation_ID:'\"+parameter+\"'}}\"  ## mutationID location\n",
    "    df01 = my_spark.read\\\n",
    "        .format(\"mongo\")\\\n",
    "        .option(\"database\",\"Mut_info\")\\\n",
    "        .option(\"pipeline\", pipeline)\\\n",
    "        .option(\"collection\",\"Mutation_Information_Merged\")\\\n",
    "        .load().limit(1)\n",
    "\n",
    "    res = df01.toJSON().collect()\n",
    "    res = json.loads(res[0])\n",
    "#     print(res)\n",
    "    tmp=list(res.keys())\n",
    "#     print(tmp)\n",
    "    return render_template('mut_inf_result.html', parameter=res, tmp = tmp, par = parameter)\n",
    "\n",
    "@app.route('/', methods=['Post'])\n",
    "def Lab_UPN():\n",
    "    Lab_UPN = request.form['u']\n",
    "#     Mutation_ID = request.form['l']\n",
    "\n",
    "#####################################\n",
    "# get data from Patient Registry data\n",
    "#     Lab_UPN = \"CO22348-0257\"\n",
    "    pipeline = \"{'$match' : {Lab_UPN:'\"+Lab_UPN+\"'}}\"\n",
    "    df01 = my_spark.read\\\n",
    "        .format(\"mongo\")\\\n",
    "        .option(\"database\",\"Patient_Registry\")\\\n",
    "        .option(\"pipeline\", pipeline)\\\n",
    "        .option(\"collection\",\"Patient_Registry\")\\\n",
    "        .load()\n",
    "\n",
    "    results = df01.toJSON().collect()\n",
    "    for i in range(len(results)):\n",
    "        results[i] = json.loads(results[i])\n",
    "#####################################\n",
    "# get data from Genomics data\n",
    "# query = '(select * from Genomics_merged where Lab_UPN = \"'+Lab_UPN+'\") as t'\n",
    "    query = '(select * from genomics_merged where Lab_UPN = \"'+Lab_UPN+'\") as t'\n",
    "    # query = '(select 1) AS t'\n",
    "    df1 = my_spark.read.format('jdbc').options(driver='com.mysql.jdbc.Driver', url=jdbc_url, dbtable=query).load()\n",
    "    \n",
    "    rdd = df1.rdd\n",
    "    b = rdd.map(tuple)\n",
    "    genomics_result=b.collect()\n",
    "    \n",
    "    ###count classification\n",
    "    df1_count = df1.groupby(\"Classification\").agg({'Classification':'count'})\n",
    "    rdd = df1_count.rdd\n",
    "    b = rdd.map(tuple)\n",
    "    cla_count=b.collect()\n",
    "\n",
    "    return render_template('table.html', test = results, genomics = genomics_result, cla_count = cla_count)\n",
    "\n",
    "\n",
    "###protein variant\n",
    "@app.route('/test/Pro_var', methods = ['Post'])\n",
    "def index3():\n",
    "    Pro_var = request.form['m']\n",
    "    Trans_ID = request.form['n']\n",
    "    Trans_var = request.form['o']\n",
    "    Gene = request.form['p']\n",
    "    print(Pro_var)\n",
    "    print(Trans_ID)\n",
    "    print(Trans_var)\n",
    "    print(Gene)\n",
    "    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Protein Variant':'\"+Pro_var+\"'},{'Transcript ID':'\"+Trans_ID+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "    if len(Gene)!=0:\n",
    "        if len(Pro_var)!=0:\n",
    "            if len(Trans_ID)!=0:\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Protein Variant':'\"+Pro_var+\"'},{'Transcript ID':'\"+Trans_ID+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Protein Variant':'\"+Pro_var+\"'},{'Transcript ID':'\"+Trans_ID+\"'}]}}\"\n",
    "            else:\n",
    "                Trans_ID = 'No Transcript ID Input'\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Protein Variant':'\"+Pro_var+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Protein Variant':'\"+Pro_var+\"'}]}}\"\n",
    "        else:\n",
    "            Pro_var = 'No Protein Variant Input'\n",
    "            if len(Trans_ID)!=0:\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Transcript ID':'\"+Trans_ID+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Transcript ID':'\"+Trans_ID+\"'}]}}\"\n",
    "            else:\n",
    "                Trans_ID = 'No Transcript ID Input'\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Gene Symbol':'\"+Gene+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    pipeline = \"{'$match' : {'Gene Symbol':'\"+Gene+\"'}}\"\n",
    "                    \n",
    "    else:\n",
    "        Gene = 'No Gene Input'\n",
    "        if len(Pro_var)!=0:\n",
    "            if len(Trans_ID)!=0:\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Protein Variant':'\"+Pro_var+\"'},{'Transcript ID':'\"+Trans_ID+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Protein Variant':'\"+Pro_var+\"'},{'Transcript ID':'\"+Trans_ID+\"'}]}}\"\n",
    "            else:\n",
    "                Trans_ID = 'No Transcript ID Input'\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Protein Variant':'\"+Pro_var+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Protein Variant':'\"+Pro_var+\"'}]}}\"\n",
    "        else:\n",
    "            Pro_var = 'No Protein Variant Input'\n",
    "            if len(Trans_ID)!=0:\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Transcript ID':'\"+Trans_ID+\"'},{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    pipeline = \"{'$match' : {$and:[{'Transcript ID':'\"+Trans_ID+\"'}]}}\"\n",
    "            else:\n",
    "                Trans_ID = 'No Transcript ID Input'\n",
    "                if len(Trans_var)!=0:\n",
    "                    pipeline = \"{'$match' : {$and:[{'Transcript Variant':'\"+Trans_var+\"'}]}}\"\n",
    "                else:\n",
    "                    Trans_var = 'No Transcript Variant Input'\n",
    "                    a=\"No input\"\n",
    "        \n",
    "                \n",
    "\n",
    "    df01 = my_spark.read\\\n",
    "        .format(\"mongo\")\\\n",
    "        .option(\"database\",\"Mut_info\")\\\n",
    "        .option(\"collection\",\"Mutation_Information_Merged\")\\\n",
    "        .option(\"pipeline\", pipeline)\\\n",
    "        .load()\n",
    "    df01.show()\n",
    "    df01 = df01[[\"Mutation_ID\"]].dropDuplicates()\n",
    "    df01 = df01.na.drop(how=\"all\")\n",
    "    rdd = df01.rdd\n",
    "    b = rdd.map(list)\n",
    "    res1 = b.collect()\n",
    "    \n",
    "############\n",
    "#     print(res1)\n",
    "    res1_tuple = (i[0] for i in res1)\n",
    "#     print(1)\n",
    "#     print(res1_tuple)\n",
    "    dat = '('\n",
    "    for i in res1_tuple:\n",
    "        dat+=\"'\"\n",
    "        dat+=i\n",
    "        dat+=\"'\"\n",
    "        dat+=','\n",
    "    dat = dat[:-1]\n",
    "#     print(2)\n",
    "    dat+=')'\n",
    "    query = '(select * from genomics_merged where Mutation_ID in '+dat+') as t'\n",
    "#     print(query)\n",
    "    df1 = my_spark.read.format('jdbc').options(driver='com.mysql.jdbc.Driver', url=jdbc_url, dbtable=query).load()\n",
    "#     print(3)\n",
    "    rdd = df1.rdd\n",
    "    b = rdd.map(tuple)\n",
    "    genomics_result=b.collect()\n",
    "#     print(4)\n",
    "    ###count classification\n",
    "    df1_count = df1.groupby(\"Classification\").agg({'Classification':'count'})\n",
    "    rdd = df1_count.rdd\n",
    "    b = rdd.map(tuple)\n",
    "    cla_count=b.collect()\n",
    "#     print(5)\n",
    "    \n",
    "    \n",
    " #######Patient_Registry####   \n",
    "    Lab_UPN_list = [i[7] for i in genomics_result]\n",
    "    Lab_UPN_list_str = '['\n",
    "    for i in Lab_UPN_list:\n",
    "        Lab_UPN_list_str += \"'\"\n",
    "        Lab_UPN_list_str += i\n",
    "        Lab_UPN_list_str +=\"'\"\n",
    "        Lab_UPN_list_str +=','\n",
    "    Lab_UPN_list_str = Lab_UPN_list_str[:-1]\n",
    "    Lab_UPN_list_str += ']'\n",
    "\n",
    "\n",
    "    pipeline = \"{'$match' : {'Lab_UPN':{'$in':\"+Lab_UPN_list_str+\"}}}\"\n",
    "\n",
    "    df01 = my_spark.read\\\n",
    "        .format(\"mongo\")\\\n",
    "        .option(\"database\",\"Patient_Registry\")\\\n",
    "        .option(\"pipeline\", pipeline)\\\n",
    "        .option(\"collection\",\"Patient_Registry\")\\\n",
    "        .load()\n",
    "####### res2 : patient information    \n",
    "    res2 = df01.toJSON().collect()\n",
    "    res2 = [json.loads(i) for i in res2]\n",
    "\n",
    "    for i in res2:\n",
    "        del i[\"_id\"]\n",
    "        \n",
    "    a=[]\n",
    "    for i in res2:\n",
    "        for j in i[\"history\"]:\n",
    "            a.append(i[\"history\"][j]['type'])\n",
    "\n",
    "    history_count = [[x,a.count(x)] for x in set(a)]\n",
    "\n",
    "    history_list = list(set(a))\n",
    "    \n",
    "    \n",
    "    return render_template('pro_var.html', Gene = Gene, parameter=res1, mutation_count=len(res1), Pro_var = Pro_var, Trans_var = Trans_var, Trans_ID = Trans_ID, genomics = genomics_result, genomics_count = len(genomics_result), cla_count = cla_count, history_count = history_count, history_list = history_list, patient_inf = res2, patient_number = len(res2))    \n",
    "\n",
    "# @app.route('/table/')\n",
    "# def index2():\n",
    "#     return render_template('table.html', test = results)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    app.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
